src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataReaderExtensions.cs

```diff
@@ -171,11 +173,58 @@ private static void ReadCollectionItem(ODataReader reader, Stack<ODataItemBase>
                     }
 
                     break;
+                case ODataReaderState.DeltaResourceSetStart:
+                    ODataDeltaResourceSet deltaResourceSet = (ODataDeltaResourceSet)reader.Item;
+                    Contract.Assert(deltaResourceSet != null, "ResourceSet should never be null.");
```

Comment:
Given that the statement above this one does a cast, i.e., `(ODataDeltaResourceSet)reader.Item` and not type conversion, i.e., `reader.Item as ODataDeltaResourceSet`, which scenarios would give rised to `deltaResourceSet` variable being null?
Doesn't the case statement guarantee that we have a `DeltaResourceSet` in our hands? Is this assert statement necessary?

Reply:
I still think here the case statement doesn't guarantee that we have a `DeltaResourceSet`, the only thing the case statement tells is that we are currently reading a `DeltaResourceSet`. The item could be a valid DeltaResourceSet or a null or etc. I need  to confirm this though